### PR TITLE
ranger: check the collation when checking like func

### LINKED ISSUE #15988: LIKE operator malfunctions for COLLATE 'latin1_bin'
Consider the following statements:

```sql
CREATE TABLE t0(c0 INT, c1 TEXT(10), PRIMARY KEY(c0));
CREATE INDEX i0 ON t0(c1(10));
INSERT INTO t0(c0, c1) VALUES (0, NULL);
SELECT * FROM t0 WHERE t0.c1 LIKE '@%' COLLATE 'latin1_bin'; -- expected: {}, actual: {0, NULL}
```

Although the predicate should evaluate to `NULL`, the `SELECT` fetches a row. That the predicate should evaluate to `NULL` is demonstrated by the following statement:

```sql
SELECT * FROM t0 WHERE (t0.c1 LIKE '@%' COLLATE 'latin1_bin') IS NULL; -- expected: {0, NULL}
```
Environment:
```
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                          |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-160-g82529c886
Git Commit Hash: 82529c886cb1cfff7810476a136acf81284bac6e
Git Branch: master
UTC Build Time: 2020-04-01 11:44:27
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #15988  

Problem Summary:

We didn't check the LIKE func's collation when checking whether it can be used to build range.

### What is changed and how it works?

What's Changed:

Check the collation info, and if the LIKE's collation info is different with the column's.
Don't use it to build range.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### CODE REVIEW

File: util/ranger/checker.go [hunk 1]
```diff
@@ -98,6 +98,10 @@ func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction
 }
 
 func (c *conditionChecker) checkLikeFunc(scalar *expression.ScalarFunction) bool {
+	_, collation, _ := scalar.CharsetAndCollation(scalar.GetCtx())
```
Comment (bb7133) [inline util/ranger/checker.go:101]: LGTM for this check(according to your explanations in the PR description).

But I don't get the point why we got a full index scan for `LIKE`, if the collations are not compatible?

**Comment by mrigger** [comment]
Any updates on this? After this is fixed, I can continue testing collations.

**Comment by bb7133** [comment]
LGTM

**Comment by zz-jason** [comment]
@winoros do we need to cherry-pick this PR to release branches.

**Pull request merged**